[SPARK-9708] [MESOS] Spark should create local temporary directories in Mesos sandbox when launched with Mesos

Zariel opened the pull request:
This is my own original work and I license this to the project under the project's open source license

Commit:
when running in mesos use the provided scratch space

AmplabJenkins commented:
Can one of the admins verify this patch?

dragos commented:
I doubt this works with dynamic allocation under Mesos. It also completely shuns the previous configuration parameters. Here's what docs say:

> spark.local.dir	Directory to use for "scratch" space in Spark, including map output files and RDDs that get stored on disk. This should be on a fast, local disk in your system. It can also be a comma-separated list of multiple directories on different disks. NOTE: In Spark 1.0 and later this will be overriden by SPARK_LOCAL_DIRS (Standalone, Mesos) or LOCAL_DIRS (YARN) environment variables set by the cluster manager.

andrewor14 commented:
ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41879 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41879/consoleFull) for   PR 8358 at commit [`30fe90b`](https://github.com/apache/spark/commit/30fe90bfcc71bd74ce7ab44d07900297cba43e42).

SparkQA commented:
[Test build #41879 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41879/console) for   PR 8358 at commit [`30fe90b`](https://github.com/apache/spark/commit/30fe90bfcc71bd74ce7ab44d07900297cba43e42).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41879/
Test PASSed.

tnachen commented:
@dragos what previous parameters are you referring to? SPARK_EXECUTOR_DIRS?
I think this logic actually makes sense, as we (Mesos) set the MESOS_DIRECTORY whenever we launch a Mesos executor to point to its sandbox, which ideally should be where all the local temporary data is written to with a Mesos Task as it can be later GCed. If this is the driver in client mode, then it should default back to wherever it is.
The consideration I think is the order of evaluation of these parameters I suppose, but I'm thinking if YARN is already overriding all other options I don't think it's far off for Mesos to also do the same.

dragos commented:
@tnachen I copy-pasted the Spark configuration docs in my previous comments. Here's what I meant by other parameters though:

- `spark.local.dir` Directory to use for "scratch" space in Spark
- ` SPARK_LOCAL_DIRS` - In Spark 1.0 and later this will override the pervious one

So, yes, the order of evaluation is one of the problems. Is there a way to configure `MESOS_DIRECTORY` itself? People may want to put *multiple* directories on different volumes, for performance. Since `MESOS_DIRECTORY` is just one, this won't be possible anymore.

It's different in Yarn because Yarn already has this concept: it gives you *several* scratch directories, precisely for this purpose.

The other (big) problem is dynamic allocation w/ external shuffle service (but we can test and only use `MESOS_DIRECTORY` when the shuffle service is disabled). See my comments on [SPARK-9708](https://issues.apache.org/jira/browse/SPARK-9708)

Zariel commented:
I dont think there is an issue with running Mesos and wanting to run over multiple disks, as it is the responsibility of whoever is managing Mesos to setup this space, running Spark on Mesos should essentially be a black box. Mesos recommends that to use multiple Mesos work dirs one should use something like raid or lvm.

`MESOS_DIRECTORY` is a sandbox managed by Mesos itself, its is derived from the framework id, executor and task and the Mesos `WORK_DIR` option which gets set when Mesos launches a task.

As currently implemented `SPARK_LOCAL_DIRS` is the lowest priority and in the if block, being overridden by both YARN and `SPARK_EXECUTOR_DIRS`, I think it would make more sense to open a separate issue to solve this and make sure that it does the correct thing with regards to the docs (or update the docs).

Regarding the dynamic allocation issues this can be changed to be disabled when dynamic allocation is enabled, or a config option can be added to explicitly enable this.

dragos commented:
I think we should respect `SPARK_LOCAL_DIRS`, but fallback to `MESOS_DIRECTORY` as a default (instead of `java.io.tmpdir`, as it currently stands).

Commit:
Disable mesos sandbox with dynamic allocation

Dont use the Mesos provided sandbox when dynamic allocation is
used to prevent the shuffle files from being GC'd by Mesos.

Fix priority order so that SPARK_LOCAL_DIRS is used above all else
and fall back to MESOS_DIRECTORY and other places.

Zariel commented:
What should be the correct order for the local dirs? Currently as I can see the priority is `YARN_LOCAL_DIRS > LOCAL_DIRS > SPARK_EXECUTOR_DIRS > SPARK_LOCAL_DIRS > spark.local.dir > java.io.tmpdir`

To me it makes sense that when running in Yarn or Mesos Spark should use their local space if it is available, if someone is running applications in Mesos then its fair to assume they (cluster operator) can provide sufficient local disk space and performance.

Ive updated the PR to disable the Mesos sandbox when dynamic allocation is enabled, and also adjusted the local dir priroity so that it now looks like this

`YARN_LOCAL_DIRS > LOCAL_DIRS > SPARK_LOCAL_DIRS > SPARK_EXECUTOR_DIRS > MESOS_DIRECTORY > spark.local.dir > java.io.tmpdir`

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42372 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42372/consoleFull) for   PR 8358 at commit [`00cf541`](https://github.com/apache/spark/commit/00cf541063a6b78c73987c24e9801d63a0f2ca04).

SparkQA commented:
[Test build #42372 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42372/console) for   PR 8358 at commit [`00cf541`](https://github.com/apache/spark/commit/00cf541063a6b78c73987c24e9801d63a0f2ca04).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42372/
Test FAILed.

Commit:
fix scala style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42376 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42376/consoleFull) for   PR 8358 at commit [`536903e`](https://github.com/apache/spark/commit/536903e2a2ad38811591f9249f5058ede7ffa24c).

SparkQA commented:
[Test build #42376 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42376/console) for   PR 8358 at commit [`536903e`](https://github.com/apache/spark/commit/536903e2a2ad38811591f9249f5058ede7ffa24c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42376/
Test PASSed.

tnachen commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -655,15 +655,19 @@ private[spark] object Utils extends Logging {
     } else if (conf.getenv("SPARK_EXECUTOR_DIRS") != null) {
       conf.getenv("SPARK_EXECUTOR_DIRS").split(File.pathSeparator)
+    } else if (conf.getenv("MESOS_DIRECTORY") != null
+      && !conf.getBoolean("spark.dynamicAllocation.enabled", false)) {
```
we should check spark.shuffle.service.enabled instead

tnachen commented:
@andrewor14 besides the settings check I think it LGTM

Commit:
check spark.shuffle.service.enabled

Zariel commented:
@tnachen done

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42492 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42492/consoleFull) for   PR 8358 at commit [`280a4d7`](https://github.com/apache/spark/commit/280a4d7dceb37b945e1664ddfed8ff45e7dfbd6a).

vanzin commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -655,15 +655,19 @@ private[spark] object Utils extends Logging {
       // created the directories already, and that they are secured so that only the
       // user has access to them.
       getYarnLocalDirs(conf).split(",")
+    } else if (conf.getenv("SPARK_LOCAL_DIRS") != null) {
```
my 2 bits: this should stay where it was, or at least be below `SPARK_EXECUTOR_DIRS`, which is the standalone equivalent of `getYarnLocalDirs()`.

tnachen replied:
The origin complication was the shuffle service, but I think perhaps we should just keep it below as @vanzin mentioned since we added the check now. I also think that It's inconsistent to have the env variable take precendence where the spark.local.dir property doesn't.

tnachen replied:
@dragos do you know if people are currently using multiple directories on different volumes?
IMO we should probably try to make that possible via Mesos especially with persistent volumes introduced already, and not relying on users having a set place set with SPARK_LOCAL_DIRS.

dragos replied:
I agree with @vanzin.

Here's the order of precedence that I think makes most sense

- `SPARK_LOCAL_DIRS` environment variable
- `spark.local.dirs` configuration
- `MESOS_DIRECTORY` environment variable
- `java.io.tmpdir`

This way, if people don't care about the local directory they fallback to Mesos dir while in Mesos and `/tmp` outside. But if people *do* care, they can configure it. Seems to be the best of both worlds.

Thoughts?

andrewor14 replied:
@dragos Yes, makes sense. @Zariel can you move the `SPARK_LOCAL_DIRS` below `SPARK_EXECUTOR_DIRS` as others have suggested?

Zariel replied:
What about the YARN options? Shall I leave that as the most preferable if running inside YARN?

SparkQA commented:
[Test build #42492 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42492/console) for   PR 8358 at commit [`280a4d7`](https://github.com/apache/spark/commit/280a4d7dceb37b945e1664ddfed8ff45e7dfbd6a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42492/
Test FAILed.

tnachen commented:
retest please

tnachen commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43095 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43095/consoleFull) for   PR 8358 at commit [`280a4d7`](https://github.com/apache/spark/commit/280a4d7dceb37b945e1664ddfed8ff45e7dfbd6a).

SparkQA commented:
[Test build #43095 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43095/console) for   PR 8358 at commit [`280a4d7`](https://github.com/apache/spark/commit/280a4d7dceb37b945e1664ddfed8ff45e7dfbd6a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43095/
Test PASSed.

Zariel commented:
@tnachen is there anything my end which needs to be done to merge this?

tnachen commented:
@Zariel I think we need to document this logic somewhere in the docs, but otherwise I think we should merge this and make changes afterwards if needed. And we'll need a actually Spark committer to review and merge this anyway :) (@andrewor14)

andrewor14 commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -655,15 +655,19 @@ private[spark] object Utils extends Logging {
     } else if (conf.getenv("SPARK_EXECUTOR_DIRS") != null) {
       conf.getenv("SPARK_EXECUTOR_DIRS").split(File.pathSeparator)
+    } else if (conf.getenv("MESOS_DIRECTORY") != null
+      && !conf.getBoolean("spark.shuffle.service.enabled", false)) {
```
style nit, put `&&` on previous line

andrewor14 commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -655,15 +655,19 @@ private[spark] object Utils extends Logging {
       conf.getenv("SPARK_EXECUTOR_DIRS").split(File.pathSeparator)
+    } else if (conf.getenv("MESOS_DIRECTORY") != null
+      && !conf.getBoolean("spark.shuffle.service.enabled", false)) {
+      // running inside Mesos and have a local directory to use
```
This needs to be expanded:
```
// Mesos already creates a directory per Mesos task. Spark should use that directory
// instead so all temporary files are automatically cleaned up when the Mesos task ends.
// Note that we don't want this if the shuffle service is enabled because we want to
// continue to serve shuffle files after the executors that wrote them have already exited.

```

andrewor14 commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -655,15 +655,19 @@ private[spark] object Utils extends Logging {
-      Option(conf.getenv("SPARK_LOCAL_DIRS"))
-        .getOrElse(conf.get("spark.local.dir", System.getProperty("java.io.tmpdir")))
-        .split(",")
+      conf.get("spark.local.dir", System.getProperty("java.io.tmpdir")).split(",")
```
actually, we can rewrite this whole thing as:
```
// In non-YARN mode ...
Option(conf.getenv("SPARK_EXECUTOR_DIRS")).map(_.split(File.pathSeparator))
  .orElse(Option(conf.getenv("SPARK_LOCAL_DIRS")).map(_.split(",")))
  // Mesos already creates a directory per ...
  .orElse(Option(conf.getenv("MESOS_DIRECTORY")).map(_.toSeq))
  .orElse(conf.getOption("spark.local.dir").map(_.split(",")))
  .getOrElse(sys.props("java.io.tmpdir").split(","))
  .toArray
```
I think the precedence is a little clearer this way

andrewor14 commented:
@Zariel This looks pretty good. Should we log a message or something if we detect shuffle service AND `MESOS_DIRECTORY`? Otherwise it might be confusing to the user why his application is using the mesos dir sometimes but not others.

Zariel commented:
@andrewor14 that sounds like a good idea, log it at INFO or WARN?

Commit:
review comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #43922 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43922/consoleFull)** for PR 8358 at commit [`fa5a0bb`](https://github.com/apache/spark/commit/fa5a0bb3e405f09febddb6eb93c0f8612ed2ec16).

SparkQA commented:
**[Test build #43922 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43922/consoleFull)** for PR 8358 at commit [`fa5a0bb`](https://github.com/apache/spark/commit/fa5a0bb3e405f09febddb6eb93c0f8612ed2ec16).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43922/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #43924 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43924/consoleFull)** for PR 8358 at commit [`546af5a`](https://github.com/apache/spark/commit/546af5a98c135f145d1a2f1369de5587b121a125).

SparkQA commented:
**[Test build #43924 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43924/consoleFull)** for PR 8358 at commit [`546af5a`](https://github.com/apache/spark/commit/546af5a98c135f145d1a2f1369de5587b121a125).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43924/
Test PASSed.

andrewor14 commented:
Yup, this looks good. I'm merging it into master. Thanks for your work @Zariel.
